[MultiRep HEIC] Ensure data is preserved when copying rich content

pxlcoder opened the pull request:
#### 1d46d6999740b7e33536645aadfdc0a45afce20e
<pre>
[MultiRep HEIC] Ensure data is preserved when copying rich content
<a href="https://bugs.webkit.org/show_bug.cgi?id=269348">https://bugs.webkit.org/show_bug.cgi?id=269348</a>
<a href="https://rdar.apple.com/122932418">rdar://122932418</a>

Reviewed by Richard Robinson.

* Source/WebCore/editing/cocoa/AttributedString.h:
* Source/WebCore/editing/cocoa/AttributedString.mm:

Decompose the attachment into individual bitmaps for serialization.

(WebCore::toNSObject):
(WebCore::extractValue):
* Source/WebCore/editing/cocoa/HTMLConverter.mm:

Add a HEIC attachment to the attributed string when a multi-representation
HEIC is encountered.

(HTMLConverter::_addMultiRepresentationHEICAttachmentForImageElement):
(HTMLConverter::_processElement):
(attachmentForElement):
(WebCore::editingAttributedString):
* Source/WebCore/editing/cocoa/TextAttachmentForSerialization.h:
* Source/WebCore/html/HTMLImageElement.cpp:
(WebCore::HTMLImageElement::isMultiRepresentationHEIC const):
* Source/WebCore/html/HTMLImageElement.h:
* Source/WebCore/platform/graphics/ImageAdapter.h:
* Source/WebCore/platform/graphics/mac/ImageAdapterMac.mm:
(WebCore::ImageAdapter::multiRepresentationHEIC):
(WebCore::ImageAdapter::invalidate):
* Source/WebCore/platform/ios/UIFoundationSoftLink.h:
* Source/WebCore/platform/ios/UIFoundationSoftLink.mm:
* Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/274690@main">https://commits.webkit.org/274690@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2d79b437e68f36b4060901be1a2077d224d1dc18

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/39741 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/18720 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/42097 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/42276 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/35643 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/42048 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/21625 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/16049 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/33154 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/40315 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/15829 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/34376 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/13684 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/13702 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/35331 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/43553 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/36119 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/35666 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/39446 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/14556 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/11984 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/37722 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/16159 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/16223 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5231 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/15819 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebCore/editing/cocoa/AttributedString.mm:
```diff
@@ -334,6 +347,12 @@
         return { { { RetainPtr { (NSPresentationIntent *)value } } } };
+#if ENABLE(MULTI_REPRESENTATION_HEIC)
+    if ([value isKindOfClass:PlatformWebMultiRepresentationHEICAttachment]) {
+        auto attachment = static_cast<WebMultiRepresentationHEICAttachment *>(value);
```
shouldn’t this use the dynamic objc cast thing we have?

pxlcoder replied:
Can't for soft-linked classes.

rr-codes commented on Source/WebCore/editing/cocoa/HTMLConverter.mm:
```diff
@@ -1198,6 +1207,34 @@ static Class _WebMessageDocumentClass()
+
+    NSUInteger textLength = [_attrStr length];
+
+    RetainPtr<NSString> string = adoptNS([[NSString alloc] initWithFormat:@"%C", static_cast<unichar>(NSAttachmentCharacter)]);
```
what’s the difference between a unichar and a unsigned char?

rr-codes replied:
also can this use the convenience initializer `stringWithFormat` instead?

pxlcoder replied:
It's an `unsigned short`, and is specifically used for [UTF-16 code units](https://developer.apple.com/documentation/foundation/unichar?language=objc).

This initializer is already used elsewhere in the file.

rr-codes commented on Source/WebCore/editing/cocoa/HTMLConverter.mm:
```diff
@@ -1198,6 +1207,34 @@ static Class _WebMessageDocumentClass()
+
+    [_attrStr addAttribute:NSAttachmentAttributeName value:attachment range:rangeToReplace];
+
+    _flags.isSoft = NO;
```
what do es this mean?

pxlcoder replied:
It ensures leading spaces are not suppressed after an attachment. It's already used this way for regular attachments.

rr-codes commented on Source/WebCore/editing/cocoa/HTMLConverter.mm:
```diff
@@ -2316,9 +2357,23 @@ AttributedString editingAttributedString(const SimpleRange& range, IncludeImages
+                auto fileWrapper = fileWrapperForElement(*imageElement);
+                auto attachment = adoptNS([[NSTextAttachment alloc] initWithFileWrapper:fileWrapper.get()]);
+                [string appendAttributedString:[NSAttributedString attributedStringWithAttachment:attachment.get()]];
+            }
```
can this maybe be refactored into its own method with early returns to ease readability?

rr-codes commented on Source/WebCore/html/HTMLImageElement.cpp:
```diff
@@ -917,6 +921,20 @@ bool HTMLImageElement::isSystemPreviewImage() const
+        return false;
+
+    auto& typeAttribute = m_sourceElement->attributeWithoutSynchronization(typeAttr);
+    if (typeAttribute != MULTI_REPRESENTATION_HEIC_MIME_TYPE_STRING)
```
can this just be return typeAttribute == MULTI_REPRESENTATION_HEIC_MIME_TYPE_STRING ?

pxlcoder replied:
Yup.

cdumez commented on Source/WebCore/editing/cocoa/HTMLConverter.mm:
```diff
@@ -2316,9 +2357,23 @@ AttributedString editingAttributedString(const SimpleRange& range, IncludeImages
+
+            if (!addedImage) {
+                auto fileWrapper = fileWrapperForElement(*imageElement);
+                auto attachment = adoptNS([[NSTextAttachment alloc] initWithFileWrapper:fileWrapper.get()]);
```
Can we please not hide smart pointers behind `auto`?

pxlcoder replied:
Will change.

rr-codes commented on Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.serialization.in:
```diff
@@ -94,7 +94,12 @@ header: <WebCore/ResourceRequest.h>
 
 [Nested] struct WebCore::AttributedString::AttributeValue {
+#if ENABLE(MULTI_REPRESENTATION_HEIC)
+    std::variant<double, String, URL, Ref<WebCore::Font>, Vector<String>, Vector<double>, WebCore::AttributedString::ParagraphStyleWithTableAndListIDs, RetainPtr<NSPresentationIntent>, RetainPtr<NSShadow>, RetainPtr<NSDate>, WebCore::AttributedString::ColorFromCGColor, WebCore::AttributedString::ColorFromPlatformColor, WebCore::TextAttachmentFileWrapper, WebCore::TextAttachmentMissingImage, WebCo ...
```
wow, what a variant lol

aprotyas reviewed (commented):
LGTM

pxlcoder committed:
[MultiRep HEIC] Ensure data is preserved when copying rich content
https://bugs.webkit.org/show_bug.cgi?id=269348
rdar://122932418

Reviewed by Richard Robinson.

* Source/WebCore/editing/cocoa/AttributedString.h:
* Source/WebCore/editing/cocoa/AttributedString.mm:

Decompose the attachment into individual bitmaps for serialization.

(WebCore::toNSObject):
(WebCore::extractValue):
* Source/WebCore/editing/cocoa/HTMLConverter.mm:

Add a HEIC attachment to the attributed string when a multi-representation
HEIC is encountered.

(HTMLConverter::_addMultiRepresentationHEICAttachmentForImageElement):
(HTMLConverter::_processElement):
(attachmentForElement):
(WebCore::editingAttributedString):
* Source/WebCore/editing/cocoa/TextAttachmentForSerialization.h:
* Source/WebCore/html/HTMLImageElement.cpp:
(WebCore::HTMLImageElement::isMultiRepresentationHEIC const):
* Source/WebCore/html/HTMLImageElement.h:
* Source/WebCore/platform/graphics/ImageAdapter.h:
* Source/WebCore/platform/graphics/mac/ImageAdapterMac.mm:
(WebCore::ImageAdapter::multiRepresentationHEIC):
(WebCore::ImageAdapter::invalidate):
* Source/WebCore/platform/ios/UIFoundationSoftLink.h:
* Source/WebCore/platform/ios/UIFoundationSoftLink.mm:
* Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/274690@main
